lib/src/Base/Func/SymbolicParserExprTk.cxx

```diff
@@ -94,64 +95,100 @@ Point SymbolicParserExprTk::operator()(const Point & inP) const
+        std::copy(&input_(i, 0), &input_(i, 0) + inputDimension, evaluation_.threadStack_[threadIndex].begin());
 
-        std::copy(inputStack_.begin() + inputDimension, inputStack_.end(), &result(i, 0));
         for (UnsignedInteger outputIndex = 0; outputIndex < outputDimension; ++ outputIndex)
         {
-          if (!SpecFunc::IsNormal(result(i, outputIndex)))
-            throw InternalException(HERE) << "Cannot evaluate " << formulas_[0] << " at " << inputVariablesNames_.__str__() << "=" << Point(inS[i]).__str__();
+          const Scalar value = evaluation_.threadExpressions_[threadIndex][outputIndex]->value();
+          // ExprTk does not throw on domain/division errors
+          if (evaluation_.checkOutput_ && !SpecFunc::IsNormal(value))
+            throw InternalException(HERE) << "Cannot evaluate " << evaluation_.formulas_[outputIndex] << " at " << evaluation_.inputVariablesNames_.__str__() << "=" << Point(input_[i]).__str__();
+          output_(i, outputIndex) = value;
         }
-      } // i
-    } // checkOutput_
+      }
+    }
     else
     {
       // Single formula
-      for (UnsignedInteger i = 0; i < size; ++i)
+      for (UnsignedInteger i = r.begin(); i != r.end(); ++ i)
       {
-        std::copy(&inS(i, 0), &inS(i, 0) + inputDimension, inputStack_.begin());
+        std::copy(&input_(i, 0), &input_(i, 0) + inputDimension, evaluation_.threadStack_[threadIndex].begin());
         // Evaluate expression
-        (void) expressions_[0]->value();
+        (void) evaluation_.threadExpressions_[threadIndex][0]->value();
+
+        std::copy(evaluation_.threadStack_[threadIndex].begin() + inputDimension, evaluation_.threadStack_[threadIndex].end(), &output_(i, 0));
 
-        std::copy(inputStack_.begin() + inputDimension, inputStack_.end(), &result(i, 0));
+        if (evaluation_.checkOutput_)
+          for (UnsignedInteger outputIndex = 0; outputIndex < outputDimension; ++ outputIndex)
+          {
+            if (!SpecFunc::IsNormal(output_(i, outputIndex)))
+              throw InternalException(HERE) << "Cannot evaluate " << evaluation_.formulas_[0] << " at " << evaluation_.inputVariablesNames_.__str__() << "=" << Point(input_[i]).__str__();
+          }
       } // i
-    } // !checkOutput_
+    }
+  }
+
+}; /* end struct SymbolicParserExprTkPolicy */
+
+
+Sample SymbolicParserExprTk::operator() (const Sample & inS) const
+{
+  const UnsignedInteger inputDimension = inputVariablesNames_.getSize();
+  if (inS.getDimension() != inputDimension)
+    throw InvalidArgumentException(HERE) << "Error: invalid input dimension (" << inS.getDimension() << ") expected " << inputDimension;
+  const UnsignedInteger outputDimension(outputVariablesNames_.getSize() > 0 ? outputVariablesNames_.getSize() : formulas_.getSize());
+  if (outputDimension == 0) return Sample(inS.getSize(), 0);
+  const UnsignedInteger size = inS.getSize();
+  Sample result(size, outputDimension);
+  if (size < 1e3)
```

Comment:
Hard-coded threshold, arrgggghhhhh! More seriously, it should be something defined in ResourceMap and made available in the class through an attribute. I expect it to depend on the size of the sample as well as the complexity of the formula(s), so it should be user-tunable.

Reply:
Dont worry I know, I spent this morning tuning the allocation of expressions per-thread.

Reply:
It should be a bit better now.